### DISCUSSION TITLE
testing help

### DISCUSSION THREAD

**Discussion by michelkluger** [reclosedev/requests-cache#916] [General]
I am using a CachedSession and its method get, for an external endpoint,

in my test this endpoint is not reached, so I want to mock the response. 

which I am trying to do with,

`
mock_cached_session.get.return_value.status_code = 200
`

`
mock_cached_session.get.return_value.json.return_value = [dict(mock_response)]
`
but after mocking, I am trying to test that caching time is respected, which it is when I manually test it, but not when testing.

I the mock negatively impacting the caching? how can I write such test?

**Comment by JWCook** [comment]
Are you mocking `CachedSession.get`? If so, that completely replaces that method with a `Mock` object, so there won't be any caching. There are a few options to use both mocking and caching:
* Mock `requests.Session.send()`. You can do this using [`patch.object()`](https://docs.python.org/3/library/unittest.mock.html#patch-object). You may need to add a few attributes to the return value so it behaves like a `requests.Response` object.
* Use a dedicated request mocking library. There are a few extra steps to take here, since both `requests-cache` and the mocking library will modify or extend `requests.Session`. There are [examples in the docs](https://requests-cache.readthedocs.io/en/stable/user_guide/compatibility.html) for using both [requests-mock](https://github.com/jamielennox/requests-mock) and [responses](https://github.com/getsentry/responses). Some minimal tests for these can be found under [tests/compat](https://github.com/requests-cache/requests-cache/tree/main/tests/compat).

**Reply by michelkluger** [reply]
@JWCook thank you very much, I am a bit beginner in the topic,

I wrote a test snippet,

@pytest.fixture
def patch_times() -> list[datetime]:
    TWO_MINUTES = 2 * 60
    t0 = datetime.now()
    t1 = t0 + timedelta(seconds=TWO_MINUTES)
    t2 = t0 + timedelta(seconds=20 * TWO_MINUTES)
    patch_times = [t0, t1, t2]
    return patch_times

@patch("my_module.CachedSession.get")
def test_cached_session(patch_times: list[datetime]) -> None:
    url = "http://test.com"
    for patch_time in patch_times:
        with patch("time.time", return_value=patch_time.timestamp()):
            session = CachedSession()
            response = session.get(url)
            if patch_time == patch_times[1]:
                assert response.from_cache
            else:
                assert not response.from_cache

this url does not exist, how can I mock the send here? or make this test work without adding another dependency to my project?
Thank you very much for your support <3

**Reply by JWCook** [reply]
There's an example in the requests-cache tests [here](https://github.com/requests-cache/requests-cache/blob/c2bb4b72305a9ee6ebba2690ca9a6b22311f8e8a/tests/integration/base_cache_test.py#L230) of mocking `Session.send()`. Almost all the other tests use `requests-mock` (for unit tests) or httpbin (for integration tests).

Is there a reason you don't want to add testing-only dependencies to your project? If you define these separate from your main dependencies, then users of your project won't need to install them, only other developers who want to run the test suite. A few examples:
* Poetry:  you can use dependency groups like [`tool.poetry.group.test.dependencies`](https://python-poetry.org/docs/managing-dependencies#dependency-groups)
* PDM: you can use [`tool.pdm.dev-dependencies`](https://pdm-project.org/latest/usage/dependency/#add-development-only-dependencies) 
* Plain setuptools + pip: you can use a separate `requirements-dev.txt` file

Also, since you're mocking the system time, you might find the [time-machine](https://github.com/adamchainz/time-machine) library useful ([more info here](https://adamj.eu/tech/2020/06/03/introducing-time-machine/#the-problem) on how it's different from `unittest.mock.patch()`).
